### DISCUSSION TITLE
Keep langchaingo lightweight enough for everyone

### DISCUSSION THREAD

**Discussion by edocevol** [tmc/langchaingo#264] [General]
Honestly speaking, I think we should ensure that this package is lightweight enough for everyone to use in production environments. We can implement different implementations based on the interfaces defined in this package, and these implementations should be as independent as possible from the current package. All langchain components should also follow this principle.

**Comment by tmc** [comment]
I completely agree. We've allowed a few too many third party dependencies to creep in -- can folks please file issues or leave comments here on places where we should be programming to interfaces instead so that we can drop third party dependencies from the main tree? (happy to put those impl-specific implementation modules into the tree with separate go.mod files).

As a sidebar, we do have github discussions enabled: https://github.com/tmc/langchaingo/discussions

**Reply by FluffyKebab** [reply]
We could do this with "go.starlark.net/starlark" in the math chain and tool for example. I also think things like https://github.com/tmc/langchaingo/blob/main/tools/sqldatabase/postgresql/ should also be in its own module.

**Comment by urjitbhatia** [comment]
In the spirit of above: https://github.com/tmc/langchaingo/discussions/325#discussion-5750896 :)

**Comment by jenglong1899** [comment]
I've heard some people say that the official version of langchain is overly encapsulated. I'm not very familiar with the code of our project, does our design avoid over-encapsulation?
